### DISCUSSION TITLE
[HarvesterMod] A few QoL features

### DISCUSSION THREAD

**Discussion by Pyromanc3r** [VintageModsReforged/HarvesterMod#1] [Ideas]
I have a question that is not related to this mod. Is it possible to port a function like the "reap" mod (I saw it in your repository) to version 1.5.2? More precisely, harvesting with RMB? Could you port this function?

I would also like the feature from the "[Fast Leaf Decay](https://www.curseforge.com/minecraft/mc-mods/fast-leaf-decay)", but I think it would be too much of an impudence to ask for it :)

**Comment by CrossVas** [comment]
Harvest on right click is something I am missing myself, but I think that I'll port it as a different mod. GraviSuite should stay IC2 oriented.

As for the FastLeafDecay, I'll have to wait until I understand how to transform classes or do some crazy algorithm :D

**Comment by Pyromanc3r** [comment]
Something similar to Fast Decay was implemented in the [Treecapitator](https://www.curseforge.com/minecraft/mc-mods/treecapitator) mod. Although, there the foliage disappeared instantly. Maybe you should look at its code.
In my opinion, that mod is no longer needed, because your GraviSuite has a "tree capitator mode", which I really like :)

**Reply by CrossVas** [reply]
https://github.com/CrossVas/HarvesterMod/releases/tag/1.0.0

**Reply by CrossVas** [reply]
> Something similar to Fast Decay was implemented in the [Treecapitator](https://www.curseforge.com/minecraft/mc-mods/treecapitator) mod. Although, there the foliage disappeared instantly. Maybe you should look at its code.

To be honest I don't like how TreeCapitator handles this (and yes, I've checked the code). I'll look into it later.

**Comment by Pyromanc3r** [comment]
Wow that was really fast! thanks :) I'll check out that mod later.
As for integration with other mods, only [Natura](https://www.curseforge.com/minecraft/mc-mods/natura/files/all?page=1&pageSize=50&version=1.5.2) and [XyCraft](https://eu2.contabostorage.com/f9f5ea7cee6e4089be6fd576f7af4a70:mcmodarchive/index.html#Mods/X/xycraft/1.5.2/) come to my mind. HarvestCraft seems to have a same feature for its crops (at least on newer versions of the game for sure). 
I don't know of any other mods that would add more crops.

**Reply by CrossVas** [reply]
Natura has only barley that needs compat as far as I can see.
As for XyCraft, these are multiblock crops and similar to RedPower flax I have to look into it deeper. Don't promise support for these.

**Comment by Pyromanc3r** [comment]
This is super! The mod works great :)
Maybe you should just reduce the chance of drop seeds to ~50% or less.
Right now it is guaranteed 100% drop rate. Maybe it's worth adding a config file.

About XyCraft, you don't have to worry about it and don't waste your time on it. I don't use that mod anyway. Just remembered there were some plants in there.

PS: new repository for retro MC? Nice :D Then I have a few more ideas for other mods.

**Reply by CrossVas** [reply]
Yeah, fixed seed drop in dev already. Now the seed amount is shrunk by 1 as it should be.
![image](https://github.com/user-attachments/assets/5e8b1d00-0d1d-4aa7-84d2-0ce5ba0c6f55)
Here is an example. 7 farm blocks, 2 rows. 
On the left is the result for manual harvesting, on the right - using right click harvesting. 
I think adding a config for this is redundant.

**Reply by CrossVas** [reply]
> PS: new repository for retro MC? Nice :D Then I have a few more ideas for other mods.

Be aware that I don't plan to work on this full time nor I'm going to implement everything you suggest.
It was made to keep mods organized because I'm also working on other modern versions at the same time.

**Reply by Pyromanc3r** [reply]
I understand. Thank you. I think the idea has been realised here and the topic can be closed. 
Oh and don't forget to move VintageHopper to your new repo.

**Comment by CrossVas** [comment]
@Pyromanc3r next patch :3
Need to test this on server first.


https://github.com/user-attachments/assets/58f980bd-1ad1-4523-a154-30fc461e1884

**Comment by CrossVas** [comment]
https://github.com/VintageModsReforged/HarvesterMod/releases

**Comment by Pyromanc3r** [comment]
I've been testing the Harvester mod (on server and single player) and found minor flaws.

1. these trees leave leaves after they're cut down:
(vanilla) jungle 2x2 big trees (~2-4 blocks leaves)
(BoP) palm (all leaves in air)
(BoP) bamboo (all leaves in air)
(BoP) pine (topmost 1 block)
(thaumcraft) `Greatwood` and `Silverwood` (all leaves in air). I think that's the most important part, as harvesting these saplings is a pain.

2. TreeCapitator mode not working with `Binnie Core` trees ([Binnie mods](https://eu2.contabostorage.com/f9f5ea7cee6e4089be6fd576f7af4a70:mcmodarchive/index.html#Mods/B/binnie%20mods/1.5.2/)).
Same with Thaumcraft trees. But this easily fix via MineTweaker adding logs to `oreDict`.
But Binnie already have ore dictionary `logWood`.

3. `Forestry` leaves not supported fast decay.

4. And I found a new crop. [ChocoCraft](https://eu2.contabostorage.com/f9f5ea7cee6e4089be6fd576f7af4a70:mcmodarchive/index.html#Mods/C/chococraft/1.5.2/) mod - Gysahl seeds.

**Reply by CrossVas** [reply]
I'll take a look. Thanks.

**Reply by CrossVas** [reply]
1. About leaves:
- jungle tree: this is something I can't really fix, because the main decay logic isn't handled by `Harvest Mod`, but by vanilla;
- palm trees: palm tree has a very specific pattern for leaves (see image 1) with no adjacent block, this is something I can't really fix; 
- bamboo trees: same as above, I can't really fix, a workaround would be to cause a block update on any leave block, then the algorithm will do the rest. It seems like bamboo stem block isn't providing a block update when harvested? Idk really...
- pine tree: same as jungle tree;
- thaumcraft trees: I'm not really sure if magic trees are supposed to decay that easy, but once I decide, I'll add support for it too...
2. About TreeCapitator:
- Binnie mods: I also check for leaves blocks and their metadata values. Will check and add specific support for that.
3. Same as Binnie Mods.
4. Will be added in the next patch.
UPD: Forestry, Binnie and Thaumcraft don't use `BlockLeavesBase` for their leaves. LeafDecay won't work unfortunately. TreeCapitator maybe.
![image](https://github.com/user-attachments/assets/df421dab-0327-4f08-a25b-e68b6db7adb9)

Small note: next time please use `Issue` section for such things, so I can keep easy track of it.

**Reply by CrossVas** [reply]
Redirected to #3

**Reply by CrossVas** [reply]
@Pyromanc3r, available in the next patch.
Thaumcraft:

https://github.com/user-attachments/assets/e22e5706-38c7-4b94-90c0-3900d2da851c

Forestry/Binnie's Mods.

https://github.com/user-attachments/assets/2f6ef573-062a-42dc-ad5a-be4b55f44f2f


Do note that I perform a raw check for modded leaves, meaning any decoration with these leaves are going to be harvested.

**Comment by Pyromanc3r** [comment]
Don't worry about those trees, they are insignificant trifles. Hardly anyone will even notice little things like leaves from a palm/pine trees. Thanks for supporting thaumcraft, I've been wishing for this for a long time :)


I'll disable this setting in my modpack though:
`harvester.cfg` -> `B:TreeCApitator=false`
I would just like to have TreeCapitator support only for chainsaw (for balance purposes).

>  Do note that I perform a raw check for modded leaves, meaning any decoration with these leaves are going to be harvested.

I'm not quite sure what you mean. You mean decorations like vines, moss, some flowers on the foliage? Then I don't think it's a problem.

**Reply by CrossVas** [reply]
By `raw check` I mean that I only check for block in general, unlike vanilla leaves, where I'm also checking metadata to detect real trees. 
> I would just like to have TreeCapitator support only for chainsaw (for balance purposes).

Uh, yeah, that would require me to add special support on `GraviSuite` side also, which I completely forgot about...

**Comment by CrossVas** [comment]
At this point it's getting hard to keep track of every feature reforged mods provide, especially when it's a duplicate feature among several mods.
It's like Achilles Heel of modding...

**Comment by Pyromanc3r** [comment]
I suggest then not to go too far :D Just leave the optimal option that can be controlled. Vanilla and TreeCapitator mode are already a lot. At this stage you have already rewritten the `TreeCapitator` mod from the author `bspkrs`. It's already done and all that's left is to look for bugs.

If anything needs more support than anything else, I think it's `Forestry` because of its Multifarms. Although, I haven't built them very often, but it seems that when a Multifarm cuts down a tree, the remaining foliage prevents a new tree from growing. Ideally with Fast Leaf Decay this problem should be eliminated (I'll have to check it out).

Anyway, don't go crazy and leave something you can easily control :)

**Comment by Pyromanc3r** [comment]
I just thought why don't you upload your mods to CurseForge and Modrinth? Maybe for a couple of people it would be a nice find too. Maybe someone still plays old versions of minecraft. Plus more beta testers :D
